src/dstruct/factor_graph/variable.h

```diff
@@ -39,6 +39,9 @@ namespace dd{
     // n_start_i_tally is the start position for the variable values in the array
     long n_start_i_tally;
 
+    std::vector<VariableValue> *domain; // domain of multinomial variable
```

Comment:
I hope to get rid of this, but if I strongly feel we should unwrap this into a plain `VariableValue[]` array since we already keep track of the `cardinality` and the vector is just unnecessary wrapping.